### PR TITLE
[Tutorial]NLP Sequence to sequence model for translation

### PR DESCRIPTION
**Pull request by siju-samuel**
- [x] Seq2Seq encoder/decoder model tutorial
- [x] Timestep support for LSTM Cells

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from others in the community.

### CODE REVIEW

**Comment by siju-samuel** [comment]
@kazum @srkreddy1238 @masahi please have one round of review. thanks

**Comment by kazum** [comment]
@siju-samuel, sorry for my late response. I think I can take a look this weekend.

File: nnvm/python/nnvm/frontend/keras.py [hunk 1]
```diff
@@ -131,6 +131,14 @@ def _convert_dense(insym, keras_layer, symtab):
     if keras_layer.use_bias:
         params['use_bias'] = True
         params['bias'] = symtab.new_const(weightList[1])
+    input_shape = keras_layer.input_shape
+    input_dim = len(input_shape)
+    # In case of RNN dense, input shape will be (1, 1, n)
+    if input_dim > 2:
+        input_shape = tuple(dim if dim else 1 for dim in _as_list(input_shape)[0])
+        if input_dim != 3 and input_shape[0] != input_shape[1] != 1:
```
Comment (kazum) [inline nnvm/python/nnvm/frontend/keras.py]: I think this check should be
```suggestion
        if input_dim != 3 or input_shape[0] != 1 or input_shape[1] != 1:
```
or
```suggestion
        if not (input_dim == 3 and input_shape[0] == input_shape[1] == 1):
```

File: nnvm/python/nnvm/frontend/keras.py [hunk 2]
```diff
@@ -408,27 +418,31 @@ def _convert_lstm(insym, keras_layer, symtab):
...
-    in_gate = _convert_recurrent_activation(gates[0], keras_layer)
-    in_transform = _convert_recurrent_activation(gates[1], keras_layer)
-    next_c = in_transform * in_state_c + in_gate * _convert_activation(gates[2], keras_layer, None)
-    out_gate = _convert_recurrent_activation(gates[3], keras_layer)
-    next_h = out_gate * _convert_activation(next_c, keras_layer, None)
+    time_steps = inp_shape[1]
+    in_data = _sym.squeeze(in_data, axis=0)
+    in_data = _sym.split(in_data, indices_or_sections=time_steps, axis=0)
+    for step in range(time_steps):
+        ixh1 = _sym.dense(in_data[step], kernel_wt, use_bias=False, units=units)
```
Comment (kazum) [inline nnvm/python/nnvm/frontend/keras.py]: I think no need to use range here.
```
    for step in in_data:
        ixh1 = _sym.dense(step, kernel_wt, use_bias=False, units=units)
```

File: nnvm/python/nnvm/frontend/keras.py [hunk 3]
```diff
@@ -666,12 +680,13 @@ def from_keras(model):
...
                 for n_idx, t_idx, layer in zip_node:
                     if isinstance(layer, keras.engine.InputLayer):
-                        sym = symtab.get_var(layer.name, must_contain=True)
+                        sym_name = layer.name
                     else:
                         sym_name = layer.name + ':' + str(n_idx) + ':' + str(t_idx)
-                        sym = symtab.get_var(sym_name, must_contain=True)
-                    insym.append(sym)
-
+                    # In some models, sym_name may not be available in inbound_nodes
```
Comment (kazum) [inline nnvm/python/nnvm/frontend/keras.py]: Can you share an example model where `sim_name` is not available in inbound_nodes?

File: tutorials/nnvm/nlp/keras_s2s_translate.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+# Download required files
+# -----------------------
+# Download files listed below from dmlc web-data repo.
+model_file = "s2s_translate.h5"
+data_file = "fra-eng.txt"
+
+# Base location for model related files.
+repo_base = 'https://github.com/dmlc/web-data/raw/master/keras/models/s2s_translate/'
+model_url = os.path.join(repo_base, model_file)
+data_url = os.path.join(repo_base, data_file)
```
Comment (kazum) [inline tutorials/nnvm/nlp/keras_s2s_translate.py]: The model and data files are not found in the repository yet.  Can you share them?

File: tutorials/nnvm/nlp/keras_s2s_translate.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+            break
+    return decoded_sentence
+
+######################################################################
+# Run the model
+# -------------
+# Randonly take some text and translate
+for seq_index in range(100):
+    # Take one sequence and try to decode.
+    index = random.randint(1, num_samples)
```
Comment (PariksheetPinjari909) [inline tutorials/nnvm/nlp/keras_s2s_translate.py]: If the model is trained on num_samples, i will suggest to test the model on validation dataset.

File: tutorials/nnvm/nlp/keras_s2s_translate.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+model_url = os.path.join(repo_base, model_file)
+data_url = os.path.join(repo_base, data_file)
+
+# Download files listed below.
+from mxnet.gluon.utils import download
+download(model_url, model_file)
+download(data_url, model_file)
+
+latent_dim = 256  # Latent dimensionality of the encoding space.
+num_samples = 10000  # Number of samples to train on.
```
Comment (PariksheetPinjari909) [inline tutorials/nnvm/nlp/keras_s2s_translate.py]: In the script, pretrained model is used, no training is done. Can you update the comment.

File: tutorials/nnvm/nlp/keras_s2s_translate.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+    (we keep the last LSTM state and discard the outputs).
+- A decoder LSTM is trained to turn the target sequences into
+    the same sequence but offset by one timestep in the future,
+    a training process called "teacher forcing" in this context.
+    Is uses as initial state the state vectors from the encoder.
+    Effectively, the decoder learns to generate `targets[t+1...]`
+    given `targets[...t]`, conditioned on the input sequence.
+
+This script loads the s2s.h5 model saved in repository
+https://github.com/dmlc/web-data/raw/master/keras/models/s2s_translate/lstm_seq2seq.py
```
Comment (slyubomirsky) [inline tutorials/nnvm/nlp/keras_s2s_translate.py]: This link appears to be broken

File: nnvm/python/nnvm/frontend/keras.py (hunk 3 above)
Comment (kazum) [reply nnvm/python/nnvm/frontend/keras.py]: This change ignores the encoder outputs in the decoder model, and uses zeros as an initial state instead.  It looks wrong to me.

The root cause is that you are processing layers which are included in the imported model but not relevant to the current model.  You can skip such layers with the below code.
```
    if not model._node_key(keras_layer, node_idx) in model._network_nodes:
        continue
````
Note that model._network_nodes contains keys of all nodes relevant to the current model.

**Comment by siju-samuel** [comment]
@kazum @PariksheetPinjari909 could you please review once again. Thanks.

File: nnvm/python/nnvm/frontend/keras.py [hunk 8]
```diff
@@ -131,6 +131,14 @@ def _convert_dense(insym, keras_layer, symtab):
     if keras_layer.use_bias:
         params['use_bias'] = True
         params['bias'] = symtab.new_const(weightList[1])
+    input_shape = keras_layer.input_shape
+    input_dim = len(input_shape)
+    # In case of RNN dense, input shape will be (1, 1, n)
```
[the remark this answers is not in the document]
Comment (siju-samuel) [reply nnvm/python/nnvm/frontend/keras.py]: yes. input can be (1, m, n)
Currently it support dense input of (m, n),

**Comment by merrymercy** [comment]
@kazum  @PariksheetPinjari909 Could you conclude on this pr?

**Comment by kazum** [comment]
The current version almost looks good to me.  I'll approve if my comment (https://github.com/dmlc/tvm/pull/1815#discussion_r232501118) is addressed

**Comment by PariksheetPinjari909** [comment]
seq2seq

I just ran the model for 10 test samples from 10000 to 10050. In the result the output sequence doesn't match with the sequence provided in dataset. @siju-samuel can you have a look at this.

**Comment by siju-samuel** [comment]
It wont match exactly same as with the actual translation as shown in the text file. This is a only a porting of keras s2s to tvm. so this will match exactly same with keras output. with tvm, accuracy wont improve further.

**Comment by merrymercy** [comment]
Should merge https://github.com/dmlc/web-data/pull/124 @tqchen

**Comment by tqchen** [comment]
@merrymercy dmlc/web-data#124 is merged
